thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/common/restclient/AbstractRestClient.java

```diff
@@ -0,0 +1,269 @@
+   * @param <T>  the type parameter defined as the return type of the response parser method
+   * @param url the http url
+   * @param host the host to connect to
+   * @param connectTimeout timeout in milliseconds
+   * @param readTimeout timeout in milliseconds
+   */
+  public <T> T doGet(String url, String host, Map<String, String> reqHeaders, int connectTimeout, int readTimeout,
+      ParseResponseFunction<InputStream, T> responseParserFunc) throws IOException {
+    Map<String, String> headers = new HashMap<>(reqHeaders);
+    headers.put("User-Agent", getClass().getName());
+    headers.put("Host", host);
+    headers.put("Accept", "application/json");
+
+    return doRequest(url, "GET", new byte[0], null, connectTimeout, readTimeout, headers,
+        responseParserFunc);
+  }
+
+  /**
+   * Perform a POST request to the given URL, with a JSON or raw string as content.
+   * If the payload is an object that will be serialized to JSON the isContentTypeJSON must be set to true
+   * @param <T>  the type parameter defined as the return type of the response parser method
+   * @param url the http url
+   * @param host the host to connect to
+   * @param postContent the post content
+   * @param isContentTypeJson flag indicating if the content is json type or not
+   */
+  public <T> T doPost(String url, String host, Map<String, String> reqHeaders, Object postContent,
+      boolean isContentTypeJson, ParseResponseFunction<InputStream, T> responseParserFunc) throws IOException {
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    OBJECT_MAPPER.writeValue(baos, postContent);
+    byte[] content = baos.toByteArray();
+    String contentType = isContentTypeJson ? "application/json" : null;
+
+    Map<String, String> headers = new HashMap<>(reqHeaders);
+    headers.put("User-Agent", getClass().getName());
+    headers.put("Host", host);
+    headers.put("Accept", "application/json");
+
+    return doRequest(url, "POST", content, contentType, 0, 0, headers,
+        responseParserFunc);
+  }
+
+  /**
+   * Send a request to the given URL with the given parameters.
+   * @param <T> the type parameter defined as the return type of the response parser method
+   * @param url server url.
+   * @param method HTTP method.
+   * @param content request content.
+   * @param contentType request content type.
+   * @param connectTimeoutMillis connection timeout.
+   * @param readTimeoutMillis read timeout.
+   * @param headers any additional request headers.
+   * @return T the result of the parser functions applied to the response content.
+   */
+  private <T> T doRequest(String url, String method, byte[] content, String contentType, int connectTimeoutMillis,
+      int readTimeoutMillis, Map<String, String> headers, ParseResponseFunction<InputStream, T> responseParserFunc)
+      throws IOException {
+    long startedMillis = System.currentTimeMillis();
+
+    HttpURLConnection conn = this.connectionFactory.openConnection(url);
```

Comment:
shall the `connection` be put into a try-with-resources block? so that it can be closed whenever there is any exception.

Reply:
no longer applies